### PR TITLE
Module Constructor

### PR DESCRIPTION
**Pull request by miloszm**
Implements module constructor with dedicated test contract and test,
also renames host_data to meta_data and makes meta_data members optional.

Implements issue #93

### CODE REVIEW

File: modules/constructor/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+#[no_mangle]
+unsafe fn increment(arg_len: u32) -> u32 {
+    uplink::wrap_transaction(arg_len, |_: ()| STATE.increment())
+}
+
+/// Expose `Constructor::init()` to the host
+#[no_mangle]
+unsafe fn init(arg_len: u32) -> u32 {
+    uplink::wrap_transaction(arg_len, |arg: u8| STATE.init(arg))
+}
```
Comment (ureeves) [inline modules/constructor/src/lib.rs:61]: I'm unsure about the `init` name. Proposals would be:

- `new`
- `ctor`
- `init`

**Review by ureeves — CHANGES_REQUESTED** [review]
I'm not sure we should return `None` on *any* piece of metadata. If metadata is unset and the contract requests it, I think it should panic - much like de-referencing a pointer to memory not owned by a program.

**Comment by miloszm** [comment]
Optionality of metadata was requested by Matteo, I noted down his request in issue #93. As far as name concerned, I am open for discussion.

File: piecrust/src/lib.rs [hunk 2]
```diff
@@ -118,7 +118,7 @@ mod types;
 mod vm;
 
 pub use error::Error;
-pub use session::Session;
+pub use session::{Session, CONTRACT_INIT_METHOD};
```
Comment (ureeves) [inline piecrust/src/lib.rs:121]: What is the reasoning for exporting this?

File: piecrust/src/session.rs [hunk 3]
```diff
@@ -116,6 +117,35 @@ impl Session {
...
+    /// the arguments provided.
+    ///
+    /// If one needs to specify the ID, [`deploy_with_id`] is available.
+    /// If one would like to *not* call the initialization method, [`deploy`] is
+    /// available.
+    ///
+    /// [`ModuleId`]: ModuleId
+    /// [`deploy_with_id`]: `Session::deploy_with_id`
+    /// [`deploy`]: `Session::deploy`
+    pub fn deploy_and_init<Arg>(
```
Comment (ureeves) [inline piecrust/src/session.rs:136]: It's probably better to modify the existing function `deploy` and `deploy_with_id` to take an `Option<Arg>`. Doing it this way implies the caller should somehow "know" that the given bytecode has a constructor,, which they are clearly not in a position to do.

File: piecrust/src/session.rs [hunk 4]
```diff
@@ -148,6 +178,42 @@ impl Session {
...
+    /// the arguments provided.
+    ///
+    /// If one would like to *not* specify the `ModuleId`, [`deploy`] is
+    /// available.
+    /// If one would like to *not* call the initialization method,
+    /// [`deploy_with_id`] is available.
+    ///
+    /// [`deploy`]: `Session::deploy`
+    /// [`deploy_with_id`]: `Session::deploy_with_id`
+    pub fn deploy_with_id_and_init<Arg>(
```
Comment (ureeves) [inline piecrust/src/session.rs:192]: Ditto here

**Review by ureeves — CHANGES_REQUESTED** [review]
Overall the implementation is good, but there's several issues I can see. I don't think we need to broaden the exposed API for this -  meaning we should avoid the proliferation of `deploy-*` calls, exporting constants (only used in tests).

There's also the problem of call history. With this implementation, an `init` call cannot be replayed, since we forget the data used in the call.

File: piecrust/tests/box.rs [hunk 5]
```diff
@@ -12,7 +12,7 @@ pub fn box_set_get() -> Result<(), Error> {
 
     let mut session = vm.genesis_session();
 
-    let id = session.deploy(module_bytecode!("box"))?;
+    let id = session.deploy(module_bytecode!("box"), None::<()>)?;
```
Comment (ureeves) [inline piecrust/tests/box.rs:15]: I didn't know one could do this on the enum's variant. Interesting!

**Review by ureeves — CHANGES_REQUESTED** [review]
I think the serialization logic can be further simplified. Below a suggestion on how to do that, following the same pattern as what `query` and `transact` do.

**Review by ureeves — APPROVED** [review]
LGTM

**Pull request merged**